### PR TITLE
riscv: Document new ELF segment type to hold memory tags in core dumps.

### PR DESCRIPTION
**Pull request by bsdjhb**

### CODE REVIEW

File: riscv.md [hunk 1]
```diff
@@ -20,6 +20,7 @@ assumed to be as specified by that base document.
 	* [Dynamic Table](#dynamic-table)
```
Comment (jrtc27) [inline riscv.md:20]: RISC-V's psABI puts new segments here with the name Program Header Table

Comment (jrtc27) [reply riscv.md:20]: (here being between TLS and Dynamic Table)

Comment (bsdjhb) [reply riscv.md:21]: I have moved it and adjusted it slightly to more closely match the style used in RISC-V.

File: riscv.md [hunk 2]
```diff
@@ -416,6 +417,46 @@ typedef struct {
...
+* `p_flags`: bitmask of capability tag MMU permissions
+
+* `p_offset`: segment file offset
+
+* `p_vaddr`: segment virtual address, same as the corresponding
+  `PT_LOAD` segment
+
+* `p_paddr`: 0
+
+* `p_filesz`: segment size in file, calculated as `p_mem_sz / (8 *
```
[the remark this answers is not in the document]
Comment (bsdjhb) [reply riscv.md:443]: I should be explicit about the alignment and sizes, yes.  In practice these segments are always page-aligned and a multiple of a page.

**Review by jrtc27 — APPROVED** [review]

File: riscv.md [hunk 3]
```diff
@@ -416,6 +417,44 @@ typedef struct {
...
+
+* `p_vaddr`: segment virtual address, must be page-aligned
+
+* `p_paddr`: 0
+
+* `p_filesz`: segment size in file, calculated as `p_memsz / (8 *
+  sizeof(void * __capability)` (eight 1-bit tags cover 8 capabilities of
+  memory)
+
+* `p_memsz`: segment size in memory, must be a multiple of the page size
```
Comment (brooksdavis) [inline riscv.md:446]: Is so you can compute the stride?

File: riscv.md [hunk 4]
```diff
@@ -416,6 +417,44 @@ typedef struct {
...
+
+* `p_align`: 0
+
+The tags are stored in the core file at `p_offset` as eight 1-bit tags
+in a byte. With a capability size of 16 bytes, a 4K page requires 32
+bytes in the core file.
+
+``p_flags`` contains `PF_R` if the segment's MMU permissions permit reading
+capability tags and `PF_W` if the segment's MMU permissions permit writing
+capability tags.
```
Comment (brooksdavis) [inline riscv.md:456]: Is this here as a way to include LOAD_CAP/STORE_CAP?

**Review by brooksdavis — COMMENTED** [review]
I assume this mirrors that was done for MTE? If so, it might be worth saying so.

File: riscv.md (hunk 3 above)
Comment (bsdjhb) [reply riscv.md:446]: Not really (the discussion we had when defining this for ARM is that the stride should be assumed for each specific PT_<arch>_MEMTAG_<type> vs having some kind of generic PT_MEMTAG).  However, the value of `p_memsz` here is sensible and also permits building tables mapping file ranges to virtual address ranges in debuggers.

File: riscv.md (hunk 4 above)
Comment (bsdjhb) [reply riscv.md:456]: Yes.  I find it useful in readelf output, e.g.:

```
> readelf -l thread.core 

Elf file type is CORE (Core file)
Entry point 0x0
There are 78 program headers, starting at offset 64

Program Headers:
  Type           Offset             VirtAddr           PhysAddr
                 FileSiz            MemSiz              Flg    Align
  NOTE           0x0000000000001150 0x0000000000000000 0x0000000000000000
                 0x00000000000072c4 0x0000000000000000  R      0x4
  LOAD           0x0000000000009000 0x0000000000121000 0x0000000000000000
                 0x0000000000001000 0x0000000000001000  RW     0x1000
  LOAD           0x000000000000a000 0x0000000000131000 0x0000000000000000
                 0x0000000000001000 0x0000000000001000  RW     0x1000
  LOAD           0x000000000000b000 0x0000000040171000 0x0000000000000000
                 0x0000000000003000 0x0000000000003000  RW     0x1000
  LOAD           0x000000000000e000 0x0000000040183000 0x0000000000000000
                 0x0000000000003000 0x0000000000003000  RW     0x1000
  LOAD           0x0000000000011000 0x0000000040186000 0x0000000000000000
                 0x0000000000001000 0x0000000000001000  RW     0x1000
  LOAD           0x0000000000012000 0x0000000040187000 0x0000000000000000
                 0x0000000000009000 0x0000000000009000  RW     0x1000
  LOAD           0x000000000001b000 0x0000000040190000 0x0000000000000000
                 0x0000000000001000 0x0000000000001000  RW     0x1000
  LOAD           0x000000000001c000 0x00000000401d3000 0x0000000000000000
                 0x0000000000002000 0x0000000000002000  R      0x1000
  LOAD           0x000000000001e000 0x00000000401d5000 0x0000000000000000
                 0x0000000000001000 0x0000000000001000  RW     0x1000
  LOAD           0x000000000001f000 0x00000000401e5000 0x0000000000000000
                 0x0000000000001000 0x0000000000001000  RW     0x1000
  LOAD           0x0000000000020000 0x00000000401e6000 0x0000000000000000
                 0x0000000000014000 0x0000000000014000  RW     0x1000
  LOAD           0x0000000000034000 0x00000000403e2000 0x0000000000000000
                 0x000000000001e000 0x000000000001e000  R      0x1000
  LOAD           0x0000000000052000 0x0000000040400000 0x0000000000000000
                 0x0000000000001000 0x0000000000001000  RW     0x1000
  LOAD           0x0000000000053000 0x0000000040410000 0x0000000000000000
                 0x000000000000b000 0x000000000000b000  RW     0x1000
  LOAD           0x000000000005e000 0x000000004041b000 0x0000000000000000
                 0x0000000000441000 0x0000000000441000  RW     0x1000
  LOAD           0x000000000049f000 0x000000004085c000 0x0000000000000000
                 0x0000000000004000 0x0000000000004000  RW     0x1000
  LOAD           0x00000000004a3000 0x0000000040868000 0x0000000000000000
                 0x0000000000028000 0x0000000000028000  RW     0x1000
  LOAD           0x00000000004cb000 0x0000000040890000 0x0000000000000000
                 0x0000000000200000 0x0000000000200000  RW     0x1000
  LOAD           0x00000000006cb000 0x0000000040a90000 0x0000000000000000
                 0x0000000000009000 0x0000000000009000  RW     0x1000
  LOAD           0x00000000006d4000 0x0000000040a99000 0x0000000000000000
                 0x0000000000001000 0x0000000000001000  RW     0x1000
  LOAD           0x00000000006d5000 0x0000000040a9a000 0x0000000000000000
                 0x0000000000010000 0x0000000000010000  RW     0x1000
  LOAD           0x00000000006e5000 0x0000000040e8a000 0x0000000000000000
                 0x0000000000020000 0x0000000000020000  RW     0x1000
  LOAD           0x0000000000705000 0x000000004128a000 0x0000000000000000
                 0x0000000000020000 0x0000000000020000  RW     0x1000
  LOAD           0x0000000000725000 0x00000000412aa000 0x0000000000000000
                 0x0000000000010000 0x0000000000010000  RW     0x1000
  LOAD           0x0000000000735000 0x000000004169a000 0x0000000000000000
                 0x0000000000020000 0x0000000000020000  RW     0x1000
  LOAD           0x0000000000755000 0x0000000041a9a000 0x0000000000000000
                 0x0000000000020000 0x0000000000020000  RW     0x1000
  LOAD           0x0000000000775000 0x0000000041aba000 0x0000000000000000
                 0x0000000000200000 0x0000000000200000  RW     0x1000
  LOAD           0x0000000000975000 0x0000000041cba000 0x0000000000000000
                 0x0000000000200000 0x0000000000200000  RW     0x1000
  LOAD           0x0000000000b75000 0x0000000041eba000 0x0000000000000000
                 0x0000000000200000 0x0000000000200000  RW     0x1000
  LOAD           0x0000000000d75000 0x00000000420ba000 0x0000000000000000
                 0x0000000000200000 0x0000000000200000  RW     0x1000
  LOAD           0x0000000000f75000 0x0000000060000000 0x0000000000000000
                 0x0000000000200000 0x0000000000200000  RW     0x1000
  LOAD           0x0000000001175000 0x0000000080000000 0x0000000000000000
                 0x0000000000600000 0x0000000000600000  RW     0x1000
  LOAD           0x0000000001775000 0x00000000a0000000 0x0000000000000000
                 0x0000000000200000 0x0000000000200000  RW     0x1000
  LOAD           0x0000000001975000 0x00000000c0000000 0x0000000000000000
                 0x0000000000200000 0x0000000000200000  RW     0x1000
  LOAD           0x0000000001b75000 0x00000000e0000000 0x0000000000000000
                 0x0000000000200000 0x0000000000200000  RW     0x1000
  LOAD           0x0000000001d75000 0x0000000100000000 0x0000000000000000
                 0x0000000000200000 0x0000000000200000  RW     0x1000
  LOAD           0x0000000001f75000 0x0000ffffbff00000 0x0000000000000000
                 0x0000000000080000 0x0000000000080000  RW     0x1000
  LOAD           0x0000000001ff5000 0x0000fffffff60000 0x0000000000000000
                 0x0000000000020000 0x0000000000020000  RW     0x1000
  LOAD           0x0000000002015000 0x0000fffffffff000 0x0000000000000000
                 0x0000000000001000 0x0000000000001000  R E    0x1000
  AARCH64_MEMTAG 0x0000000002016000 0x0000000000121000 0x0000000000000000
                 0x0000000000000020 0x0000000000001000  RW     0
  AARCH64_MEMTAG 0x0000000002016020 0x0000000000131000 0x0000000000000000
                 0x0000000000000020 0x0000000000001000  RW     0
  AARCH64_MEMTAG 0x0000000002016040 0x0000000040171000 0x0000000000000000
                 0x0000000000000060 0x0000000000003000  RW     0
  AARCH64_MEMTAG 0x00000000020160a0 0x0000000040183000 0x0000000000000000
                 0x0000000000000060 0x0000000000003000  RW     0
  AARCH64_MEMTAG 0x0000000002016100 0x0000000040186000 0x0000000000000000
                 0x0000000000000020 0x0000000000001000  RW     0
  AARCH64_MEMTAG 0x0000000002016120 0x0000000040187000 0x0000000000000000
                 0x0000000000000120 0x0000000000009000  RW     0
  AARCH64_MEMTAG 0x0000000002016240 0x0000000040190000 0x0000000000000000
                 0x0000000000000020 0x0000000000001000  RW     0
  AARCH64_MEMTAG 0x0000000002016260 0x00000000401d3000 0x0000000000000000
                 0x0000000000000040 0x0000000000002000  R      0
  AARCH64_MEMTAG 0x00000000020162a0 0x00000000401d5000 0x0000000000000000
                 0x0000000000000020 0x0000000000001000  RW     0
  AARCH64_MEMTAG 0x00000000020162c0 0x00000000401e5000 0x0000000000000000
                 0x0000000000000020 0x0000000000001000  RW     0
  AARCH64_MEMTAG 0x00000000020162e0 0x00000000401e6000 0x0000000000000000
                 0x0000000000000280 0x0000000000014000  RW     0
  AARCH64_MEMTAG 0x0000000002016560 0x00000000403e2000 0x0000000000000000
                 0x00000000000003c0 0x000000000001e000  R      0
  AARCH64_MEMTAG 0x0000000002016920 0x0000000040400000 0x0000000000000000
                 0x0000000000000020 0x0000000000001000  RW     0
  AARCH64_MEMTAG 0x0000000002016940 0x0000000040410000 0x0000000000000000
                 0x0000000000000160 0x000000000000b000  RW     0
  AARCH64_MEMTAG 0x0000000002016aa0 0x000000004041b000 0x0000000000000000
                 0x0000000000008820 0x0000000000441000  RW     0
  AARCH64_MEMTAG 0x000000000201f2c0 0x000000004085c000 0x0000000000000000
                 0x0000000000000080 0x0000000000004000  RW     0
  AARCH64_MEMTAG 0x000000000201f340 0x0000000040868000 0x0000000000000000
                 0x0000000000000500 0x0000000000028000  RW     0
  AARCH64_MEMTAG 0x000000000201f840 0x0000000040890000 0x0000000000000000
                 0x0000000000004000 0x0000000000200000  RW     0
  AARCH64_MEMTAG 0x0000000002023840 0x0000000040a90000 0x0000000000000000
                 0x0000000000000120 0x0000000000009000  RW     0
  AARCH64_MEMTAG 0x0000000002023960 0x0000000040a99000 0x0000000000000000
                 0x0000000000000020 0x0000000000001000  RW     0
  AARCH64_MEMTAG 0x0000000002023980 0x0000000040a9a000 0x0000000000000000
                 0x0000000000000200 0x0000000000010000  RW     0
  AARCH64_MEMTAG 0x0000000002023b80 0x0000000040e8a000 0x0000000000000000
                 0x0000000000000400 0x0000000000020000  RW     0
  AARCH64_MEMTAG 0x0000000002023f80 0x000000004128a000 0x0000000000000000
                 0x0000000000000400 0x0000000000020000  RW     0
  AARCH64_MEMTAG 0x0000000002024380 0x00000000412aa000 0x0000000000000000
                 0x0000000000000200 0x0000000000010000  RW     0
  AARCH64_MEMTAG 0x0000000002024580 0x000000004169a000 0x0000000000000000
                 0x0000000000000400 0x0000000000020000  RW     0
  AARCH64_MEMTAG 0x0000000002024980 0x0000000041a9a000 0x0000000000000000
                 0x0000000000000400 0x0000000000020000  RW     0
  AARCH64_MEMTAG 0x0000000002024d80 0x0000000041aba000 0x0000000000000000
                 0x0000000000004000 0x0000000000200000  RW     0
  AARCH64_MEMTAG 0x0000000002028d80 0x0000000041cba000 0x0000000000000000
                 0x0000000000004000 0x0000000000200000  RW     0
  AARCH64_MEMTAG 0x000000000202cd80 0x0000000041eba000 0x0000000000000000
                 0x0000000000004000 0x0000000000200000  RW     0
  AARCH64_MEMTAG 0x0000000002030d80 0x00000000420ba000 0x0000000000000000
                 0x0000000000004000 0x0000000000200000  RW     0
  AARCH64_MEMTAG 0x0000000002034d80 0x0000000060000000 0x0000000000000000
                 0x0000000000004000 0x0000000000200000  RW     0
  AARCH64_MEMTAG 0x0000000002038d80 0x0000000080000000 0x0000000000000000
                 0x000000000000c000 0x0000000000600000  RW     0
  AARCH64_MEMTAG 0x0000000002044d80 0x00000000a0000000 0x0000000000000000
                 0x0000000000004000 0x0000000000200000  RW     0
  AARCH64_MEMTAG 0x0000000002048d80 0x00000000c0000000 0x0000000000000000
                 0x0000000000004000 0x0000000000200000  RW     0
  AARCH64_MEMTAG 0x000000000204cd80 0x00000000e0000000 0x0000000000000000
                 0x0000000000004000 0x0000000000200000  RW     0
  AARCH64_MEMTAG 0x0000000002050d80 0x0000000100000000 0x0000000000000000
                 0x0000000000004000 0x0000000000200000  RW     0
  AARCH64_MEMTAG 0x0000000002054d80 0x0000ffffbff00000 0x0000000000000000
                 0x0000000000001000 0x0000000000080000  RW     0
  AARCH64_MEMTAG 0x0000000002055d80 0x0000fffffff60000 0x0000000000000000
                 0x0000000000000400 0x0000000000020000  RW     0
 ```

Here the 'RW' tags on `AARCH64_MEMTAG` correspond to `VM_PROT_READ_CAP` and `VM_PROT_WRITE_CAP`.

(It probably goes without saying, but PT_RISCV_MEMTAG_CHERI and PT_AARCH64_MEMTAG_CHERI are identical aside from the arch-specific PT_* value).

**Comment by bsdjhb** [comment]
It does somewhat mirror MTE.  MTE doesn't set p_flags to anything as there are no equivalent MMU permissions.

**Comment by jrtc27** [comment]
Huh, so I can set the MTE tags on file-backed shared mappings and watch them non-deterministically disappear as pages get paged out and back? That's... interesting.

**Comment by jrtc27** [comment]
(and potentially have them propagate to other processes for unknown amounts of time?)

**Comment by bsdjhb** [comment]
> Huh, so I can set the MTE tags on file-backed shared mappings and watch them non-deterministically disappear as pages get paged out and back? That's... interesting.

I don't know how MTE interacts with swapping or pageouts.  For all I know the use of `PROT_MTE` on a mapping wires it?

**Comment by jrtc27** [comment]
Hmm. Linux says "Note: PROT_MTE is only supported on MAP_ANONYMOUS and RAM-based file mappings (tmpfs, memfd). Passing it to other types of mapping will result in -EINVAL returned by these system calls." But if you _don't_ specify PROT_MTE can you still store tags to those pages, they'll just not be swapped in/out and not be dumped? I'm sure that'll be fun to debug...

**Comment by jrtc27** [comment]
Ok, I see, they bolted alongside all the memory attributes that specify cacheability so now you have normal memory and normal tagged memory. I guess that works, though it feels like the wrong place to put it, that's not historically been a permissions thing, just a caching thing...

**Pull request merged**